Tomo stage for testing alignment algorithms

tfarago opened the pull request:
This PR creates a virtual tomography stage. Sample is in position [0, 0, 0] and the stack of motors is below it. Top-to-bottom order of motors:

- parallel_motor_above and orthogonal_motor_above (in the same height)
- tomo_motor
- roll_motor
- lamino_motor
- vertical_motor_below
- orthogonal_motor_below

Then, there is a dummy camera which takes this stack and produces a sphere in location that depends on the positions of the motors listed above ((0, 0, 0) means "image center"). This can be used to test the alignment algorithms.

@sarkarchandan when you start the session, viewer will open, which will show a rotating sphere, whose pose is determined by the tomographic motor stack. The `await camera.get_center()` gives you the current center of the sphere (returned coordinates are x, y, z, what they mean see below) (which is changing in the session (!) because of the `tomo.velocity = ...` command). You want to keep this center around (0, 0, 0) because that's when the sphere is in the middle of the projection. The coordinate system right now is `x = ortho`, `y = beam direction`, `z = vertical`, so: `x, _, y = await camera.get_center()` gives you the position of the sphere on the projection.

TODOS:

- [ ] decide which class goes where, namely the stage should perhaps have its own directory in devices or go into `positioners`?
- [ ] documentation
- [ ] test and document backlash (@tfarago)
- [ ] tests, @sarkarchandan please could you add some basic tests when we decide on the first checkbox? I made a fast version of projecting a sphere and this should be heavily tested against all edge cases because the goodness of the sphere projection will influence the correlation (e.g. sphere out of FOV, half out of FOV, center between pixels and so on), you basically write a test with a bunch of shapes and centers and compare `make_sphere(fast=False)` and `make_sphere(fast=True)`

## In session

```python
Welcome to Concert 0.34.0dev
# *stage* session manual

## Usage

## Notes

stage > await camera.get_center()
array([0.09393, 5.49201, 0.04785]) <Unit('millimeter')>

stage > await camera.get_center()
array([-1.86957,  5.15717,  0.05057]) <Unit('millimeter')>

stage > stage_ortho.position += 1.2 * q.mm

stage > await camera.get_center()
array([-0.1817 ,  4.37504,  0.03479]) <Unit('millimeter')>

stage > stage_ortho.position -= .2 * q.mm

stage > await camera.get_center()
array([1.12212, 5.33767, 0.04503]) <Unit('millimeter')>

stage > await camera.get_center()
array([-0.87237,  5.21206,  0.05155]) <Unit('millimeter')>

stage > exit
```

## Session stage

```python
"""# *stage* session manual

## Usage

## Notes
"""

import logging
from inspect import iscoroutinefunction
import concert
concert.require("0.34.0")

from concert.quantities import q
from concert.session.utils import ddoc, dstate, pdoc, code_of
from concert.devices.cameras.dummy import TomographyStageCamera
from concert.ext.viewers import PyQtGraphViewer

LOG = logging.getLogger(__name__)
MULTIINSTANCE = False


camera = await TomographyStageCamera(shape=(2160, 2560))
viewer = await PyQtGraphViewer(show_refresh_rate=True)

# Motors
tomo = camera.stage.tomo_motor
lamino = camera.stage.lamino_motor
roll = camera.stage.roll_motor
align_ortho = camera.stage.orthogonal_motor_above
align_par = camera.stage.parallel_motor_above
stage_ortho = camera.stage.orthogonal_motor_below
stage_vert = camera.stage.vertical_motor_below

await align_ortho.set_position(1 * q.mm)
await lamino.set_position(1 * q.deg)
await roll.set_position(0.25 * q.deg)
await tomo.set_velocity(45 * q.deg / q.s)
f = viewer(camera.stream())
```

MarcusZuber reviewed (changes requested):
I would favor, if the whole geometry would be in the motor class. You have to know the stacking order of the TomographySatge in the camera. If you add a get_center function to to motor and only project and convert to pixels in the camera one could easily add other geometries.

+1 for homogeneous coordinates!

tfarago commented:
> Regarding: "decide which class goes where, namely the stage should perhaps have its own directory in devices or go into positioners?"
> 
> I'm also heavily thinking about the `Positioners` at the moment for other reasons. I don't like the current implementation and we should maybe shortly discuss them a bit.

I will re-think this.

tfarago commented:
I added backlash to our dummy motors and updated the "stage" in such a way that the roll motor and lamino motor both have backlash of 0.1 degree. @sarkarchandan, try this:
```python
import logging
from inspect import iscoroutinefunction
import concert
concert.require("0.34.0")

from concert.quantities import q
from concert.session.utils import ddoc, dstate, pdoc, code_of
from concert.devices.cameras.dummy import TomographyStageCamera
from concert.ext.viewers import PyQtGraphViewer

LOG = logging.getLogger(__name__)
MULTIINSTANCE = False

camera = await TomographyStageCamera(shape=(2160, 2560))
viewer = await PyQtGraphViewer(show_refresh_rate=True)

# Motors
tomo = camera.stage.tomo_motor
lamino = camera.stage.lamino_motor
roll = camera.stage.roll_motor
align_ortho = camera.stage.orthogonal_motor_above
align_par = camera.stage.parallel_motor_above
stage_ortho = camera.stage.orthogonal_motor_below
stage_vert = camera.stage.vertical_motor_below

await stage_ortho.set_position(1 * q.mm)
await align_par.set_position(2 * q.mm)
await lamino.set_position(1 * q.deg)
await roll.set_position(0.25 * q.deg)
# await tomo.set_velocity(45 * q.deg / q.s)
f = viewer(camera.stream())
```

then:
```
> lamino.position += 0.1 * q.deg  # # Motor moves and the sphere shifts vertically
> lamino.position -= 0.1 * q.deg  # Nothing happens with the sphere in the viewer
> lamino.position -= 0.1 * q.deg  # Motor moves and the sphere shifts vertically in the opposite direction
```

The changes I made are purely functional so that @sarkarchandan has a more realistic base for testing his alignment algorithms. I will do the refactoring with positioners and whatnot later.

MarcusZuber commented:
To make it even more "realistic", you could also add a random positioning error. Depending on the controller/motor you either have a positioning error and then it stays at the (wrong) position, or it constantly moves around within target-window.
